LE Audio: BAP scan delegator

Thalley opened the pull request:
Heavily refactors the existing scan delegator implementation. 

The main points are to move the PA and BIS sync out from the stack, and into the application. This allows for more flexibility, and avoids numerous issues that could appear with the existing implementation (e.g. handling the scan state). 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/41183
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/51308

Thalley committed:
Bluetooth: Audio: Add Scan Delegator dependency for Broadcast Sink

Per the BAP specification, if the Broadcast Sink role is
supported, then the Scan Delegator shall also be supported.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Properly document bap.h

The bap.h header file contains several structs and defines
that are not properly documented.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Make scan delegator recv state const in API

Since the data is informative only, the struct should be
const.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Remove testing guard for scan delegator set sync state

The set sync state will be used for upper layers to set the BIS
sync state. This is due to the fact that the scan delegator cannot
automatically get this information from the ISO layer, unlike
the PA sync state.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Refactor scan delegator set sync state

The bt_bap_scan_delegator_set_sync_state has been refactored
to bt_bap_scan_delegator_set_bis_sync_state, as it will,
going forward, only be used to set the BIS sync states,
and not the PA sync state. The PA sync state will, in a future
commit, but autonousmly handled by the scan delegator
based on the PA sync callbacks.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Remove autonomous BIS sync from scan delegator

The Scan delegator will rely on upper layers calling
bt_bap_scan_delegator_set_bis_sync_state instead.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add receive_state_updated for scan delegator

Add the receive_state_updated callback for the scan delegator.
This callback will be called whenever a receive state is updated
by any means, giving the upper layers the option to always know
the latest changes.

This commit also refactors the name used for the internal
receive state (which contains additional information), for
the sake of readability.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Remove PA sync handling from scan_delegator

The scan delegator will no longer be responsible for
handling the actual synchronization of the periodic
advertisers, and will offload this to higher layers.

The reason for this, is that scanning is a global state,
and should be avoided autonousmly by the stack. The
application is much better suited for handling this.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add scan delegator broadcast code callback

Add a callback that is called when the broadcast
assistant supplies a broadcast code.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Support local add_source for scan delegator

Add support for adding a source from the scan delegator itself.
This is useful if e.g. the broadcast sink syncs to a PA.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Move requested_bis_sync from recv_state to internal

The requested_bis_sync value is not part of the BASS receive
state. Moving the field makes it possible to use the
recv_state struct in more situations.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BASS server sync test case

Add a test case that tests the scan delegator and
broadcast assistant features when the scan delegator
syncs to a PA without being asked to by the
broadcast assistant.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Fix issue with bt_bap_scan_delegator_remove_source

The bt_bap_scan_delegator_remove_source failed due to invalid length,
as the buffer supplied to scan_delegator_rem_src should not
contain the opcode.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BSIM test for scan delegator BIS sync state

Add BSIM testing for bt_bap_scan_delegator_set_bis_sync_state.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Fix bug in bis_syncs_unique_or_no_pref

The check did not work when aggregated_bis_syncs
was 0. Also modfied the function to be more readable.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Fix issue with subgroup metadata for scan delegator

The scan delegator attempted to put the subgroup pointer into the
metadata data, instead of the metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Scan delegator mod source added

Add function for the scan delegator to modify the metadata
of a receive state locally.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add testing of scan delegator mod source

Add testing fo the scan delegator mod source functionality.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add cmd_bap_scan_delegator_add_src

Add command for the scan delegator to add a source
locally based on a PA sync.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add cmd_bap_scan_delegator_mod_src

Add cmd_bap_scan_delegator_mod_src to modify a BASS
receive state source as the scan delegator in the
shell.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add cmd_bap_scan_delegator_rem_src

Add the cmd_bap_scan_delegator_rem_src to remove
receive state sources locally as the Scan Delegator.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Remove enc_state from scan_delegator_set_bis_sync_state

The field is added/modified by bt_bap_scan_delegator_add_src
and bt_bap_scan_delegator_mod_src. This makes it easier
to modify the BIG sync state, without worrying about
the encryption state.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add Scan Delegator receive state iterate functions

Add bt_bap_scan_delegator_foreach_state and
bt_bap_scan_delegator_find_state to support finding
specific receive states.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Integrate scan deletagor and broadcast sink

This integrates the Scan Delegator functionality with the
Broadcast Sink functionality. The Broadcast Sink will
automatically update the receive state for the PA and BIG
sync values, based on state of the Broadcast Sink.

Similarly, a request to terminate the BIG or PA sync from
a Broadcast Assistant will terminate the Broadcast Sinks'
PA or BIG syncs.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add broadsink sink create

Add support to create a broadcast sink from a PA sync, rather
than using the broadcast sink scan functions.

This allows for the scan delegator implementation to
autonoumsly add broadcast sinks.

This refactors how the broadcast sink uses flags, to
ensure thread safetyness.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add bis_sync_req scan delegator callback

Add callback to notify the application about which
BIS it should sync to when requested by the Broadcast
Asssitant.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
samples: Bluetooth: Add Scan Delegator support to Broadcast Audio Sink

Add support for the Scan Delegator for the Broadcast
Audio Sink sample. This include advertising connectable
as well as handling PAST.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Known issues should be fixed now, and review comments from the walk-through should also have been addressed. 

It has been rebased on https://github.com/zephyrproject-rtos/zephyr/pull/51897

Thalley committed:
Bluetooth: Audio: Shell: BAP scan delegator term sync cmd

Added command to terminate the PA sync as the scan delegator.
This can also be used to cancel any pending PA syncs.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley committed:
Bluetooth: Audio: Add BAP prefix for BASS tests

add the `bap_` prefix for the BAP BASS tests.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

szymon-czapracki commented on subsys/bluetooth/audio/shell/bap_scan_delegator.c:
```diff
@@ -18,60 +18,735 @@
+	if (past_preference &&
+	    state->past_avail &&
+	    state->conn != NULL) {
+		shell_info(sh, "Syncing with PAST");
+		err = pa_sync_past(state->conn, state, state->pa_interval);
+
+		if (err != 0) {
```
```suggestion
		shell_info(sh, "Syncing with PAST");
		
		err = pa_sync_past(state->conn, state, state->pa_interval);
		if (err != 0) {
```
Wouldn't this be better structure?

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -287,6 +307,25 @@ static void test_bass_discover(void)
+		WAIT_FOR_FLAG(flag_recv_state_read);
+	}
+
+	printk("Receive state read complete\n");
```
Isn't it better to use LOG() instead of printk?

szymon-czapracki replied:
Applies in multiple places.

Thalley replied:
We don't have a log registered for our babblesim tests (yet). Maybe we should consider that

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -287,6 +307,25 @@ static void test_bass_discover(void)
+		err = bt_bap_broadcast_assistant_read_recv_state(default_conn, i);
+		if (err != 0) {
+			FAIL("Failed to read receive state with idx %u: %d\n",
+			     i, err);
+			return;
```
```suggestion
			     i, err);

			return;
```

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -428,15 +467,15 @@ static void test_bass_remove_source(void)
 	if (err != 0) {
 		FAIL("Bluetooth enable failed (err %d)\n", err);
-		return;
+		return err;
```
```suggestion

		return err;
```

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/51679#discussion_r1156943149

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -447,7 +486,7 @@ static void test_main(void)
 	if (err != 0) {
 		FAIL("Scanning failed to start (err %d)\n", err);
-		return;
+		return err;
```
```suggestion

		return err;
```

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/51679#discussion_r1156943149

We could modify it, but that would be a separate change for the entire file. Let me know if you want me to update that

szymon-czapracki replied:
This is really a nitpick, I think this can be as it is.

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -456,23 +495,95 @@ static void test_main(void)
+
+	err = common_init();
+	if (err != 0) {
+		FAIL("Bluetooth enable failed (err %d)\n", err);
+		return;
```
```suggestion
		FAIL("Bluetooth enable failed (err %d)\n", err);

		return;
```

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/pull/51679#discussion_r1156943149

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -456,23 +495,95 @@ static void test_main(void)
+
+	err = common_init();
+	if (err != 0) {
+		FAIL("Bluetooth enable failed (err %d)\n", err);
+		return;
```
```suggestion
		FAIL("Bluetooth enable failed (err %d)\n", err);

		return;
```

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/pull/51679#discussion_r1156943149

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -456,23 +495,95 @@ static void test_main(void)
+
+	err = common_init();
+	if (err != 0) {
+		FAIL("Bluetooth enable failed (err %d)\n", err);
+		return;
```
```suggestion
		FAIL("Bluetooth enable failed (err %d)\n", err);

		return;
```

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/pull/51679#discussion_r1156943149

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_scan_delegator_test.c:
```diff
@@ -5,76 +5,830 @@
+
 static volatile bool g_cb;
 static volatile bool g_pa_synced;
+CREATE_FLAG(flag_broadcast_code_received);
```
If we start to use CREATE_FLAG() isn't it a good idea to change existing "flags" in the same way? Or would this require too much work and another PR?

Thalley replied:
Good question. IMO it's not a problem to use both, even if we should ideally use `CREATE_FLAG`. I can update the test to only use `CREATE_FLAG` if you want

szymon-czapracki replied:
I think we should stick to one way of doing this just for readability, and I think it's better to do this here, as later this can be forgotten.

Thalley replied:
Added as 2 new commits

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_scan_delegator_test.c:
```diff
@@ -5,76 +5,830 @@
+
+	state = sync_state_get_by_pa(sync);
+	if (state == NULL) {
+		FAIL("Could not get sync state from PA sync %p\n", sync);
+		return;
```
```suggestion
		FAIL("Could not get sync state from PA sync %p\n", sync);
		return;
```
```suggestion
		FAIL("Could not get sync state from PA sync %p\n", sync);

		return;
```

szymon-czapracki commented on tests/bsim/bluetooth/audio/src/bap_scan_delegator_test.c:
```diff
@@ -5,76 +5,830 @@
+	const uint16_t pref_context = BT_AUDIO_CONTEXT_TYPE_CONVERSATIONAL;
+	struct bt_bap_scan_delegator_mod_src_param param;
+	uint8_t pref_context_metadata[4] = {
+		0x03,
```
I know those are just tests, but maybe it will be good to add a comment describing this 0x03 value.
I.e is this just a random number within metadata, or does this represent sth?

Thalley replied:
It is the size of the data :) It's an LTV struct (length-type-value), but I'll add a comment

szymon-czapracki commented on tests/bsim/bluetooth/audio/test_scripts/bap_bass_client_sync.sh:
```diff
@@ -4,23 +4,23 @@
 cd ${BSIM_OUT_PATH}/bin
 
-printf "\n\n======== Running BASS and BASS client test =========\n\n"
+printf "\n\n======== Running Running BASS Client Sync =========\n\n"
```
Running + Running = Ultra fast running :D

Thalley commented on subsys/bluetooth/audio/bap_scan_delegator.c:
```diff
@@ -31,12 +31,16 @@ LOG_MODULE_REGISTER(bt_bap_scan_delegator, CONFIG_BT_BAP_SCAN_DELEGATOR_LOG_LEVE
 
+enum bass_recv_state_internal_flag {
+	BASS_RECV_STATE_INTERNAL_FLAG_NOTIFY_PEND,
+
```
Intentional to separate the flags from the "FLAG_NUM". This is convention in Zephyr as far as I can tell

Thalley committed:
tests: Bluetooth: BAP Scan Delegator test only use FLAGs

Modify the BAP Scan Delegator test to only use the FLAG macros,
instead of volatile booleans, to be more similar to the other
LE Audio BSIM tests.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
tests: Bluetooth: BAP Broadcast Assistant test only use FLAGs

Modify the BAP Broadcast Assistant test to only use the FLAG macros,
instead of volatile booleans, to be more similar to the other
LE Audio BSIM tests.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -328,6 +328,18 @@ struct bt_bap_scan_delegator_cb {
+	/**
+	 * @brief Receive state updated
+	 *
+	 * @param conn       Pointer to the connection to a remove device if the
```
I make that typo way too often...
